Summercleaning2020

Technologicat opened the pull request:
Hi JuliaFEM team,

As previously discussed privately, in the spirit of open source, let's bring this refactoring work out into the open for discussion and further development.

Here's a PR containing my suggestion of how the code of `Materials.jl` could be reorganized so that it becomes easier to read, and contribute to.

**This PR is not final.** It is currently posted for discussion.


What this PR changes:

- **Modules!**
  - Using modules prevents inadvertent name conflicts between different parts of the library. Any particular material model may now define whatever names it pleases for its internal use; these won't interfere with anything defined elsewhere.
  - Names intended for export must be declared explicitly, using Julia's `export` facility. This explicitly defines the public API of the module.
  - Each material model now lives in its own module. As of this writing, we have the source files `idealplastic.jl` and `chaboche.jl`, each of which defines a module.
  - The strain/stress increment optimizers live in an `Increments` module, in source file `increments.jl`.
  - Utility functions (containing e.g. various rank-4 tensors) have been collected into a `Utilities` module, source file `utilities.jl`, so individual material models won't need to re-invent them (applying the [DRY](https://en.wikipedia.org/wiki/Don't_repeat_yourself) principle).
    - See the existing models for examples for how to use the utilities in a material model.
  - The top-level `Materials.jl` loads all these modules with `include`, and then brings them into scope by `using .SomeModule`. Then it exports those names that are considered interesting for the end user of `Materials.jl`, so that those names become available directly at the top level of the `Materials` package. The user of `Materials.jl` doesn't need to care about how the package is internally organized.
    - The load order is utilities first, then material models, then the strain/stress optimizers.
    - The re-export step is done manually in the top-level `Materials.jl`.
      - *Explicit is better than implicit*. Although the proverb is famously associated with Python, considering clarity, it holds in software engineering in general (as do many of the other principles of [The Zen of Python](https://www.python.org/dev/peps/pep-0020/)).
      - In general, an explicit static list of names is easier for developer tooling to recognize (for features such as autocompletion).
      - There is [`Reexport.jl`](https://github.com/simonster/Reexport.jl) we could use to automatically re-export all public names from a submodule, but that's one more dependency. Each abstraction is one more thing for the reader to learn (as noted by Paul Graham, in [On Lisp](http://paulgraham.com/onlisp.html)). But if that becomes desirable, we can easily switch to using `Reexport.jl`.
- **Docstrings!** Each public function now comes with a docstring explaining its purpose.
  - The target audience is the end users of `Materials.jl`, i.e. numerical developers.
  - The docstring includes the function call signature, as suggested in the [Julia style guide](https://docs.julialang.org/en/v1/manual/style-guide/index.html). The format is "whatever gets the important points across" followed by "make it look like the rest of the Julia docs". For example, we avoid repeating the parameter list, which is already in the signature. Parameters whose purpose is obvious from the name and type alone often need no further documentation. We have attempted to mention anything that is not obvious from context.
  - Some docstrings still need work. Grep for TODO comments.
- **Comments!** *Code tells you how, comments tell you why* ([Jeff Atwood](https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/)).
  - If a piece of documentation is intended for the caller of the function, it's in the docstring. If it's only needed for understanding the internal details, then it's in a comment.
  - Some calculations now have clarifying comments describing what the aim is, or what some particular variable means.
  - Some parts that could be improved have been marked with a TODO comment explaining how.
- Use of **whitespace**.
  - Follow typical whitespace rules for source code, such as `x = 42`, `f(a=2, b=3)`, `a + b`, etc. Deviate only when doing so improves readability, especially with a mathematical audience in mind (`a*b + c`).
  - Grouping suggests structure. Have a blank line between "paragraphs", like in prose, e.g. at the point where the "topic" of a calculation changes. [*Code is prose*](https://matthewbischoff.com/code-is-prose/).
  - Make sure any similar parts of code are whitespaced consistently, for easy comparison.
- **Clarity** of code, in general. Such as:
  - Ease reading by being careful with variables: use descriptive names, have no unused variables, limit the scope as much as reasonably possible.
  - Define helper functions for operations such as converting between *(E, ν)* and *(μ, λ)*, or marshaling the problem state to/from a vector. This helps separate the levels of abstraction, making it easier to see at a glance what is being done at each level (expressed in a vocabulary appropriate for that level).
  - Favor an explicit `if`/`end` block instead of abusing the short-circuiting AND operator as in `(e < tol) && return ...`. It is quite rare to see the AND operator used like this.
    - The use of the OR operator in a similar manner is more common, such as in `converged(res) || error(...)`. Some even consider it idiomatic (at least in Python circles). A possible explanation is that "result converged, or [else] error" makes sense when read as English, whereas the meaning of "result converged, and return" is not immediately as obvious. The latter is naturally expressed as "if result converged, then return", which translates to an `if`/`end` block.
- Improved **type annotations**.
  - The main aim is to explicitly document assumptions regarding the types, especially those of function parameters.
  - Attention has been especially paid to give the logically correct type to each function parameter wherever reasonably possible.
    - Because a function type is [contravariant](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) in its input, the correct type is the most general one admissible (e.g. `Real` or `AbstractVector{<:Real}`). This allows *using* those functions with any concrete subtype thereof (e.g. `Float64` or `Vector{Float64}`, respectively), without requiring changes to the function if there later arises a need to switch to a different concrete subtype.
- Some specific changes:
  - Move the strain/stress optimizers into a single source file, `increments.jl`.
    - The code of each is short, and all three optimizers use the same algorithm.
    - The shared algorithm is now a higher-order function, `optimize_dstrain!`, which takes as one of its arguments a routine to actually update the `dstrain` variable in whatever way is appropriate for the particular optimization being performed. This makes each specific optimizer into just a definition of such a routine, and then a call into the general optimizer.
  - Remove the `viscoplastic` model, because the code was out of date, and the model is a special case of `chaboche`. Now `chaboche` contains a note as to how to set the material parameters to obtain the `viscoplastic` special case.
  - Clean up tests.
    - Use a consistent naming scheme for testsets.
    - Avoid numeric constants and format conversions (e.g. Voigt) in the test code proper; give them meaningful names beforehand.
    - Use `let` to avoid leaking local variables from individual test cases (because it's top-level code, not inside a function).
      - Nest `let` blocks where needed, to explicitly document the scope of variables and prevent mistakes.
    - Remove commented-out, unused test code.


As of this writing, **TODO**:

- I'd like to go over the unit tests with Reijo, to make sure they physically make sense, and improve their readability. My contributions so far have concentrated on the software engineering side of things.
  - We have already cleaned up the Chaboche shear test case; others to follow later.
- The out-of-date `examples` folder should be deleted, after stripping it for any useful parts for unit tests. Unit tests are better for documentation; because they must pass, they are more likely kept up to date as the library evolves.
- Rebase against the latest master, and resolve conflicts.

And finally, at the risk of derailing the discussion into a [bikeshed](https://en.wikipedia.org/wiki/Law_of_triviality):

**NOTE**: *as of 2020-08-07, a solution to this is currently in place. Marked as **DONE** below.*

- Since we're type-annotating things anyway, I'd like to use a generic real number type instead of hard-coding `Float64`. The PR sketches out a future where most explicit mentions of `Float64` have been banished in favor of `Real` and `<:Real` as appropriate, but as yet **it is not final**.
  - Everything should Just Work, except perhaps `ForwardDiff.jacobian` for sufficiently exotic input types; but that's not our problem anyway, it's up to the developers of `ForwardDiff` and whatever provides the particular exotic concrete subtype of `Real` the user is interested in.
  - What we *can* guarantee is that *our code* works with general `Real` numbers, and that the tests pass (including our uses of any third-party libraries, such as `ForwardDiff`, with the versions specified in `Project.toml`) when the concrete subtype is `Float64`.
  - IMHO, this is not overengineering; it is exactly the role of a modern type system (such as Julia's) to encode abstract facts (e.g. "this function expects any kind of real numbers") into the static type.
    - It's probably not currently possible to encode the fact that we expect at least 64 bits if floats are used (because in Julia, concrete types are final, `FloatN` and `FloatM` are not related when `M ≠ N`). However, since this code is for scientific users, that is the baseline assumption anyway. We can safely assume anyone trying to run the code with `Float32` knows what they're doing.
  - As the PR currently stands, the code has the serious problem that the parameterization changes the public API. So instead of `m = Chaboche(...)`, in my current draft, we have to say `m = Chaboche{Float64}(...)`. What to think of this?
    - My own immediate reaction is that that's just stupid, because `Float64` is the de facto scientific standard, and likely to remain so for the foreseeable future. So for brevity, the user should be allowed to omit specifying the type when `Float64` is what they want.
      - [`Parameters.jl`](https://github.com/mauro3/Parameters.jl), which the struct definitions use anyway, seems to have some support for defaults for type parameters, but it might not support my exact use case, or at least I haven't figured out how to spell out what I want in a way it understands.
    - On the other hand, in the event of an eventual future migration to `Float128` or something else unforeseen, would an outdated default mask bugs? The principle that *explicit is better than implicit* suggests we should require users to write `Chaboche{Float64}` already now, to guard against distant future breakage. As a user, having to do that would make me feel the library is overengineered (whether or not it actually is).
    - **DONE**: Maybe we could have a `ChabocheGeneric{T}`, and then just define `Chaboche = ChabocheGeneric{Float64}`? This way both current and future users might be happy.
    - Some other solution? Discussion welcome.

(This final point is likely one of those things that requires fewer lines to fix than to explain.)

jvaara commented on src/chaboche.jl:
```diff
@@ -1,96 +1,167 @@
+    seff_dev = dev(stress - X1 - X2)
+    # von Mises yield function
+    f = sqrt(1.5)*norm(seff_dev) - (R0 + R)  # using elastic trial problem state
+    isplastic = (f > 0.0)
```
did you plan to use this on the next line or is it a leftover from your debugging states?

Technologicat replied:
Ah, left over from debugging. Fixed. Thanks for the catch.

jvaara commented on src/chaboche.jl:
```diff
@@ -1,96 +1,167 @@
+
+        dotp = ((f >= 0.0 ? f : 0.0)/Kn)^nn  # plasticity multiplier, see equations (3) and (4) in Chaboche 2013
+        dp = dotp*dtime  # |dε_p|, using backward Euler (dotp is ∂ε_p/∂t at the end of the timestep)
+        n = sqrt(1.5)*seff_dev/norm(seff_dev)  # Chaboche: a (tensorial) unit direction, s.t. 2/3 * (n : n) = 1
```
Mathematically speaking, this is ∂f/∂σ if you wish to add that to the comment :-)

Technologicat replied:
`n = ∂f/∂σ`? Sure, I'll add it.

jvaara commented on src/chaboche.jl:
```diff
@@ -1,95 +1,163 @@
+    # resulting deviatoric plastic stress (accounting for backstresses Xm)
+    seff_dev = dev(stress - X1 - X2)
+    # von Mises yield function
+    f = sqrt(1.5)*norm(seff_dev) - (R0 + R)  # using elastic trial problem state
```
Mathematically, this is ∂f/∂σ

jvaara commented on src/chaboche.jl:
```diff
@@ -99,49 +167,89 @@ function integrate_material!(material::Chaboche)
+"""
+    create_nonlinear_system_of_equations(material::Chaboche)
+
+Create and return an instance of the equation system for the delta form of the
```
perhaps `incremental form` instead of `delta form`?

Technologicat replied:
Sure, maybe clearer. Changed.

Technologicat commented:
As for the parametric polymorphism bikeshed,

```julia
@with_kw mutable struct GenericChaboche{T <: Real} <: AbstractMaterial
    ...
end

Chaboche = GenericChaboche{Float64}
```

indeed works, and requires no extra effort from the user to use the currently most sensible real number type (`Float64`). I'll push the changes related to this once the review is complete.

jvaara commented on src/chaboche.jl:
```diff
@@ -99,49 +167,89 @@ function integrate_material!(material::Chaboche)
+
+The function `g!` is intended to be handed over to `nlsolve`.
+"""
+function create_nonlinear_system_of_equations(material::Chaboche{T}) where T <: Real
```
I started thinking that maybe this unpacking, that is done twice now, is in vain. Would it make sense to move the `g!` function definition directly to the `integrate_material!`? Only the accepted `stress` is corrupted in the `integrate_material!` due to the elastic trial, but that can be worked around easily. What do you think?

Technologicat replied:
Code is read much more often than it is written, and [reading code is harder than writing it](https://www.joelonsoftware.com/2000/04/06/things-you-should-never-do-part-i/), so almost always, clarity trumps speed.

However, numerics has the peculiar characteristic that, unlike in general software engineering, in numerics it's often possible to see *a priori* where performance bottlenecks will arise. From the viewpoint of the FEM solution process, this code obviously runs in a tight loop. Speed may be a concern here.

One thing that comes to mind though, is that Julia's compiler is fairly advanced - are we sure it won't optimize away the extra unpacking automatically? The documentation explicitly states that the compiler may inline functions wherever it sees fit, and in our case, `integrate_material!` is the only call site of the function `create_nonlinear_system_of_equations`, which in turn is private to `ChabocheModule`. It's an obvious candidate for inlining. Of course, the compiler must still respect the rules of lexical scope, so any (mutable) variables created in the callee must be kept distinct from those of the caller.

What I'm getting at is that if other properties of the material are only read (not written to), only one copy is really needed for them, and only `stress` needs to exist in duplicate. So if the function is inlined, it is in theory possible to refactor the result as unpack once, then copy the (already unpacked) stress variable, and for anything that is only read, just use a single copy. (Or even, if some property is only ever read once, just use a pointer to its memory address in the original data structure, and read it via that, no copy needed. But this is getting pretty close to the metal already.)

Still, as D. Knuth famously stated, [*Premature optimization is the root of all evil*](http://wiki.c2.com/?PrematureOptimization). So to be sure, I think we should profile this both ways to see whether we get a speed increase that is significant enough to pay for the decrease in clarity.

To be done later?

jvaara commented on src/idealplastic.jl:
```diff
@@ -1,73 +1,94 @@
+Ideal plastic material: no hardening. The elastic region remains centered on the
+origin, and retains its original size.
+"""
+function integrate_material!(material::IdealPlastic{T}) where T <: Real
```
In the English literature the "_ideal plastic_" term is actually "**perfect plastic**". The "ideal plastic" term is a direct translation from the Finnish teaching materials :-). Perhaps we should rename it to `PerfectPlastic` to ring more bells.

Technologicat replied:
Thanks, didn't know that. Let's change it.

jvaara commented on src/increments.jl:
```diff
@@ -0,0 +1,156 @@
+# higher-order function. The individual specific optimizer functions
+# (`update_dstrain!)` only need to define the "meat" of how to update `dstrain`.
+"""
+    optimize_dstrain!(material::AbstractMaterial, dstrain::AbstractVector{<:Real},
```
Again touching the terminology, in my mind, we are solving a root-finding problem, instead of optimization. Would `find_dstrain!` be more descriptive?

jvaara replied:
These methods are made to be able to easily simulate stress states produced by some of the most common test machines. For example in the push-pull machine (with smooth specimen) we know that the stress state is uniaxial (in the measuring area), so basically we have to find a strain increment that produces zero stress in the other directions.

Technologicat replied:
More descriptive, and shorter, too, so yes, let's change that.

(But as you say, it's only Newton-Raphson in the *currently implemented* strain/stress optimizers, so the wording should be kept ambiguous enough to allow for other kinds of optimization in the future. `find` is fine for that.)

Thanks for the description, that'll complement the docstrings nicely. :)

jvaara commented on src/utilities.jl:
```diff
@@ -0,0 +1,183 @@
+isotropic_elasticity_tensor(lambda::T, mu::T) where T <: Real = 3 * lambda * IV(T) + 2 * mu * IS(T)
+
+"""
+    lame(E::Real, nu::Real)
```
Perhaps add the link to the useful table found in Wikipedia:
https://en.wikipedia.org/wiki/Template:Elastic_moduli

Technologicat replied:
Yes.

jvaara commented:
So regarding this more strict type

> As for the parametric polymorphism bikeshed,
> 
> ```julia
> @with_kw mutable struct GenericChaboche{T <: Real} <: AbstractMaterial
>     ...
> end
> 
> Chaboche = GenericChaboche{Float64}
> ```
> 
> indeed works, and requires no extra effort from the user to use the currently most sensible real number type (`Float64`). I'll push the changes related to this once the review is complete.

Great!

jvaara commented:
I like the changes, and the tremendous job that was done especially in writing docstrings and commenting. I am not too worried that the code should be so future-proof that if e.g. Float128 comes around, everything would work like charm. I assume the more strict typesetting does not interfere with the `ForwardDiff` functionalities? Do you know why are the tests not running for this PR?

Technologicat commented:
Thanks for the review, and for the kind words!

I pushed the `GenericThisAndThat{T} = ...`, `ThisAndThat = GenericThisAndThat{Float64}` changes, and fixed a typoed variable name. It seems I hadn't run the tests against the very latest edit (which was done quickly in a meeting with Reijo - he suggested it would be nice to experiment with using the different algorithms in NLsolve in our Chaboche implementation). Now I did. All tests pass.

The code, as it currently stands, should already work like a charm when `Float128` eventually comes along, provided the authors of our dependencies do their part. Just use `GenericChaboche{Float128}`, and we're done. :)

The primary reason I prefer `Real` instead of a specific type is that it is the logically correct thing to do. My thoughts being on implementation details while writing the original comment may have resulted in the wrong emphasis - sorry for the confusion.

Future-proofing is a free bonus. The meat is that our code makes no assumptions about the internal representation of the real numbers it uses, requiring only that they abide by some reasonable approximation to standard real-number arithmetic. This is exactly what an abstract `Real` type stands for.

This helps make assumptions explicit, which speeds up reading the code. If a function advertises itself as requiring `Float64`, I have to worry about the possibility that it may somewhere [twiddle the bits of the floating-point representation](https://randomascii.wordpress.com/2012/01/23/stupid-float-tricks-2/) to do its thing, because that's then perfectly legal. By requiring a generic type that doesn't support such trickery, the code explicitly promises not to do that. Performance is not affected - when the compiler sees that the code is actually being called with `Float64` as the concrete type, it will generate the appropriate specialization, just as if `Float64` had been hardcoded into the signatures.

`ForwardDiff` can sometimes be finicky. For example, I got a not-very-helpful `MethodError` when trying to add a scalar to a vector term, assuming (on my part) that it would broadcast, as such things almost always do. For the quantity part itself, I suppose the code would have worked fine, but the automatic differentiation didn't like it. (In this particular case, I had made a mistake in my pen-and-paper algebra when factoring terms - turns out I didn't need to add that scalar afterall. In case you're wondering, it was about the new, cleaner expressions for the components 8-19 of `g!`. The expressions actually in the code, as of posting this PR, are correct, and the model passes its tests.)

In open-source projects, the clean way to handle things like this is to not handle them in client code. If `ForwardDiff` itself isn't behaving right, then that's where to fix the problem. If some exotic, user-made concrete subclass of `Real` doesn't quite support all the features `ForwardDiff` rightly expects a `Real` to have (so, in promising to behave like a `Real`, it's breaking [its own contract](https://en.wikipedia.org/wiki/Design_by_contract)), then that's where to fix the problem.

As long as we have any reasonable concrete subclass of `Real`, I'm fairly confident the whole program (including dependencies) will work. So I wouldn't worry about `Float128` - it's such an obvious target that our dependencies will cater for it too when it becomes relevant (provided those projects are still maintained at that time). Also, if someone is crazy enough to run our code with `Rational`, I see no reason to prevent that (applying the "we're all responsible users" principle).

As for the tests, `] activate .`, `cd("test")`, `include("runtests.jl")` works for me, and all tests pass, so I suppose you refer to the CI failure? In that case I currently have no idea, except that there are at least merge conflicts with `master` that we need to fix anyway before this PR can go in. I suggest we do that first, and then look again at what CI says at that time. That should make it easier to debug what's going on.

Technologicat commented:
I merged the changes from the latest `master` to my work branch, and pushed the updates to this PR.

Don't mind the silly commit message - I originally did this by accidentally merging in the wrong direction in my local copy, but fortunately, given that silly changeset, `git` was smart enough to apply the changes **from** master **to** `summercleaning2020`. :)

I already applied some of my refactorings to the merged code (modules, use the `SymmN` aliases and the `lame` utility function), to quickly get the tests to run (and they pass!), but some remain to be done. I've added TODO comments to that effect in `DSA.jl` and `memory.jl`. Also, I'll clean up the tests while at it.

Now that the merge conflicts are gone, we can start figuring out what's going wrong in CI...

Technologicat commented:
If I read [the report](https://travis-ci.com/github/JuliaFEM/Materials.jl/jobs/369718079) right, a test in `utilities.jl` is erroring out on Julia 1.0. We still support that? :3

Seems I have some fixing to do...

TeroFrondelius commented:
> is erroring out on Julia 1.0. We still support that?
Julia 1.0 is LTS, the next LTS will be 1.6, but it will take a year or so. I would like to support LTS version, of course it depends how big extra effort it is.

Technologicat commented:
Good point about LTS. I'll investigate.

TeroFrondelius commented:
I think we should test against 1.0 and 1.5 (and nightly as it is). Can you change it here: https://github.com/JuliaFEM/Materials.jl/blob/da0433247cad39e2639b3864b50ff29035b40b10/.travis.yml#L7

Technologicat commented:
> I think we should test against 1.0 and 1.5 (and nightly as it is). Can you change it here:
> 
> https://github.com/JuliaFEM/Materials.jl/blob/da0433247cad39e2639b3864b50ff29035b40b10/.travis.yml#L7

Sure. Updated.

Technologicat commented:
The refactoring of the `DSA` and `memory` material models is now done. The code now follows the same format as the other models.

Their tests have been cleaned up, too. Stuff local to a testset stays local, without leaking to the top level when running the tests.

This completes the merge with the latest master.

Docstrings still require some work. This needs to be done by someone familiar with these particular material models. I think we should document the meaning of each property, similarly to how this PR already explains `chaboche` in the relevant docstrings. For easy grepping, I've marked the properties of `DSA` and `memory` I currently have no idea about as `???`.

Also, the general "what use case this model is for" explanation for each material model is a bit thin at the moment. I suppose DSA stands for *dynamic strain aging*, but what are the technical implications of incorporating that phenomenon into a model? And what is the *strain memory effect* incorporated by the `memory` material model? Both seem to be different generalizations of `chaboche`; so, in what kind of particular circumstances would one use these models instead?

What I'm after is a simple, short description for a generally intelligent audience, not necessarily familiar with these models, like saying that our `chaboche` implementation incorporates two backstresses, and exhibits both isotropic and kinematic hardening.

And while at it, are there papers describing each model we could cite in the docstrings, as further reading for such an audience? (I suppose at least Chaboche 2013, which we discussed earlier, and which is already alluded to in some of the comments I've added to the code. I'll dig up the proper citation. Anything else?)

Finally, one technical thing that came to mind: does the ordering of the fields within a `@with_kw` struct matter (e.g. for [ABI](https://en.wikipedia.org/wiki/Application_binary_interface) reasons)?

To make all models consistent with each other (whence easier to read), I moved the `jacobian` of `DSAVariableState` so that it is introduced after the problem state variables proper.

From my original cover letter to this PR, the following TODOs remain:

- I'd like to go over the unit tests with Reijo, to make sure they physically make sense, and improve their readability. My contributions so far have concentrated on the software engineering side of things.
  -  We have already cleaned up the Chaboche shear test case; others to follow later.
- The out-of-date examples folder should be deleted, after stripping it for any useful parts for unit tests. Unit tests are better for documentation; because they must pass, they are more likely kept up to date as the library evolves.

Technologicat commented:
Ah, and the new TODOs:

- Fix Julia 1.0 support (build failing in `utilities.jl`)
- Finish the docstrings (requires domain knowledge)
- Cite a paper (or several, if applicable) for each model (requires domain knowledge)
- Maybe we could also update the README to include some simple usage examples (e.g. from the tests)?

Technologicat commented:
Julia 1.0 compatibility fixed in 0af8bb3.

Turns out creating a `Diagonal` by calling `I` from `LinearAlgebra` wasn't a thing in Julia 1.0. I was using that feature in a bunch of tests in `test_utilities.jl`. I was going to solve this by defining a custom `eye` function, locally for those tests, to provide it (see c092ed0). But even better, a bare `I` matches any size, and works starting from Julia 1.0. My hunch is that's more julian, too, not to specify the size when it is not needed.

Remaining TODOs:

- Strip the examples folder for useful parts, incorporate those into the automated tests, then delete the examples folder.
- Clean up automated tests with Reijo, from a physics perspective.
- Finish the docstrings (requires domain knowledge).
- Cite a paper (or several, if applicable) for each model (requires domain knowledge).
- Update README with some examples.
  - Or at least suggest to look at the automated tests for documented examples, and give a link pointing to the `test` folder?

Technologicat commented:
Ok, so I've been looking at the `examples` directory, and done some cleaning there. The motivation is that a new user of `Materials.jl` is likely to look at the `examples` directory, with the general expectation that the code distributed there runs successfully, and is relevant to learning how to use the package.

Some of the code there works, but some errors out in my Julia 1.5 environment. Particularly:

- `abstractmaterial_full_workflow.jl` errors when attempting to use `Tensors.gradient`. The error message is `ERROR: LoadError: MethodError: no method matching Float64(::ForwardDiff.Dual{Nothing,Float64,6})`.

  I've looked at [the underlying code in `ForwardDiff`](https://github.com/JuliaDiff/ForwardDiff.jl/blob/master/src/dual.jl) (mainly to figure out the role of the type parameter `T`, which in this case, for some reason, is `Nothing`) and [the docs of the higher-level `Tensors` package](https://kristofferc.github.io/Tensors.jl/stable/man/automatic_differentiation/), but neither of those made me any wiser.

  So unless the reason of the error is apparent to someone else here, perhaps it doesn't make sense to use more time on it now.

- `one_elem_disp_chaboche.jl` errors, suggesting a problem in `FEMBase`. The error message is `ERROR: LoadError: ArgumentError: Package FEMBase does not have TimerOutputs in its dependencies`.

  What I did was just `Pkg.add("JuliaFEM")`. The package installed successfully, but running the mentioned example in `Materials.jl` errors out with the message above.

  Again, unless the reason is apparent to someone else, I suggest skipping this part for now. This PR concentrates on `Materials.jl` itself, not its integration with `JuliaFEM`.

Remaining TODOs:

- Clean up automated tests with Reijo, from a physics perspective.
- Finish the docstrings (requires domain knowledge). @jvaara, could you look at this when you have time?
- Cite a paper (or several, if applicable) for each model (requires domain knowledge).
  - As the reader may have noticed, I have added citations to two papers by Chaboche. Remaining are the other models.
    - For the perfect plastic model, we could cite a textbook. **Suggestions?**
    - DSA and Memory could use citations to papers where they have been investigated, if applicable. **Suggestions?**
- Update README with some examples, or suggest looking in the automated tests for examples.

vjhaanpaa commented:
>         * DSA and Memory could use citations to papers where they have been investigated, if applicable. **Suggestions?**

`DSA.jl` is indeed a dynamic strain ageing model based on `chaboche` that I have been working on in the spring and earlier in the summer. It will ideally be gone through in more detail in an upcoming materials physics master's thesis and in an article later on, so I will add up to date citations at those points. 

I will also add comparison to `chaboche` then, since I just got back to the office and have to recap things before I put my foot in my mouth trying to compare the two from memory.

jvaara commented:
> * DSA and Memory could use citations to papers where they have been investigated, if applicable. **Suggestions?**

https://doi.org/10.1016/j.ijplas.2012.09.011 For DSA. The BiBTeX entry from google scholar is as follows:
@article{chaboche2013viscoplastic,
  title={Viscoplastic constitutive equations of combustion chamber materials including cyclic hardening and dynamic strain aging},
  author={Chaboche, J-L and Gaubert, Anais and Kanout{\'e}, P and Longuet, Arnaud and Azzouz, Farida and Mazi{\`e}re, Matthieu},
  journal={International journal of plasticity},
  volume={46},
  pages={1--22},
  year={2013},
  publisher={Elsevier}
}

https://doi.org/10.1016/0749-6419(85)90018-X could do for Memory. 
@article{nouailhas1985constitutive,
  title={On the constitutive equations for cyclic plasticity under nonproportional loading},
  author={Nouailhas, D and Chaboche, JL and Savalle, S and Cailletaud, G},
  journal={International Journal of Plasticity},
  volume={1},
  number={4},
  pages={317--330},
  year={1985},
  publisher={Elsevier}
}

jvaara commented:
> * For the perfect plastic model, we could cite a textbook. **Suggestions?**

Simo's book (Computation Inelasticity, https://www.springer.com/gp/book/9780387975207) is probably one of the better ones. Unfortunately, the used notation differs from ours somewhat. I tried to check whether the consistent tangent moduli we have implemented (I cannot recall where did copy the moduli formulation for our implementation) coincides with the one presented in the book section "_3.6.2 Consistent Elastoplastic Moduli. Perfect Plasticity_", but it is not that simple with a quick glance - they show some resemblance but I'm not 100% certain :P. The implemented tangent moduli has been tested with FE-connection in the beam examples (although it worked, it doesn't mean it retains the quadratic rate of convergence for the global Newton-Raphson). Should we adapt Simo's (or whatever textbook we choose to follow) formulation/notation for the tangent moduli?

![image](https://user-images.githubusercontent.com/1643320/90383462-dad83d00-e088-11ea-9da1-1adc8a952eb0.png)

TeroFrondelius commented:
> Ok, so I've been looking at the `examples` directory, and done some cleaning there. The motivation is that a new user of `Materials.jl` is likely to look at the `examples` directory, with the general expectation that the code distributed there runs successfully, and is relevant to learning how to use the package.

We should use Literate.jl to generate our examples/notebooks. The idea is that an example is a test and Literate.jl generates the markdown (and notebook) files from its comments. Here is an example I have made myself: https://github.com/JuliaFEM/FEMMaterials.jl/blob/master/examples/example_3dbeam.jl.

Technologicat commented:
@vjhaanpaa: Ok. Thanks! The citation @jvaara provided is fine for now. You can of course add more later.

@jvaara: Thanks. Added these citations to the docstrings.

I had actually cited `10.1016/j.ijplas.2012.09.011` for the original Chaboche model. Is there a more appropriate citation for that?

As for the consistent tangent moduli / jacobian, that's actually something I discussed with Reijo back in June. To us, it seems that it might not be consistent - some terms may be missing. It may be close enough in practice, which would explain why the FE beam solver works, but convergence is indeed not necessarily quadratic. Thanks for the screencap - I'll think about this a bit more, and return to the topic later.

If the jacobian turns out to need fixing, maybe we could do that later, in another PR? This one is mainly a refactor, so it would be nice to finalize and merge, once the remaining refactoring-related TODOs are done.

@TeroFrondelius: literate programming as in Knuth? Why not. Thanks for the `Literate.jl` example.

Yes, let's move the examples after we reformat them. We'll all sleep easier at night if the examples contain automated tests. ;)

Technologicat commented:
While on the topic of literate programming, personally, I like the SICP school. Emphasis mine:

*First, we want to establish the idea that a computer language is not just a way of getting a computer to perform operations but rather that it is a novel formal medium for expressing ideas about methodology. Thus, **programs must be written for people to read, and only incidentally for machines to execute.***

--[Abelson, Sussman & Sussman: Structure and Interpretation of Computer Programs, 2nd ed. (1996)](https://mitpress.mit.edu/sites/default/files/sicp/full-text/book/book.html), preface to the first edition.

Of course, in numerics, occasionally modulo performance concerns...

Both Knuth's ideas about LP ([summary on Wikipedia](https://en.wikipedia.org/wiki/Literate_programming)) and the first edition of SICP were published in 1984, so perhaps it was in the zeitgeist.

Technologicat commented:
We're almost done...

The available citations are in place.

The automated tests have been given a final once-over with Reijo. They should be ok now. Some minor things came up in our review, added to TODOs below.

Remaining **TODO**s:

- **Units for E and R0? MPa?**
- Docs: make it explicit that by "biaxial", we mean one normal component and one shear, because e.g. two normal components is also arguably "biaxial".
- Update strain optimizer docstrings with material from @jvaara's description earlier in this thread. [Adding this here so I won't forget to do this.]
- Finish the docstrings (requires domain knowledge).
  - If someone (@jvaara? @vjhaanpaa?) has compact descriptions for what each of the material parameters of `DSA` and `Memory` do, I could include them in the docstrings. (See anything marked `???`.)
  - But if not, I can take a look by reading the cited papers and writing the descriptions myself - but that likely won't make it in time for this PR. Of course, it can be done later, so IMHO that's not a showstopper...
- Update README:
  - Suggest looking in the automated tests for usage examples.
  - Mention what `Materials.jl` is: a material models add-on for JuliaFEM, concentrating on plasticity.

Then, I think it's time to complete this PR; any further changes (especially those not related to refactoring) can go into a new one.

**Reijo had a nice suggestion**: currently, we have a separate optimizer routine for each combination of `dstrain` components (11, and 11+12). How about unifying these?

I thought about this a bit. To do this, we could change the API so that we accept a Voigt vector of `dstrain` (or `dstress` in the stress-driven case), and instruct the user to set to `missing` any components that are unknown, to be computed by optimization. And for those components the user wants fixed, just set them to the desired numeric value.

This way we get optimizers for not only uni- and biaxial tests, but for any possible combination.

jvaara commented:
> I had actually cited `10.1016/j.ijplas.2012.09.011` for the original Chaboche model. Is there a more appropriate citation for that?

Probably this one is more suitable:
https://doi.org/10.1016/0749-6419(89)90015-6

jvaara commented:
> If the jacobian turns out to need fixing, maybe we could do that later, in another PR? This one is mainly a refactor, so it would be nice to finalize and merge, once the remaining refactoring-related TODOs are done.

Agreed.

jvaara commented on src/DSA.jl:
```diff
@@ -1,174 +1,313 @@
+`C2`, `D2`: parameters governing behavior of backstress X2
+`Q`: shift parameter for yield strength evolution
+`b`: multiplier for yield strength evolution
+`w`: ???
```
A parameter that controls the average waiting time dislocation is arrested at localized obstacles.
Represents a strain increment produced when all arrested dislocations overcome localized obstacles and move towards to the next pinned configuration. In practice, this parameter controls how fast the effective aging time reacts to plastic strain.
\dot{t}_a = 1 - t_a \dot{p} / w

jvaara commented on src/DSA.jl:
```diff
@@ -1,174 +1,313 @@
+`Q`: shift parameter for yield strength evolution
+`b`: multiplier for yield strength evolution
+`w`: ???
+`P1`, `P2`: ???
```
A parameter that controls the maximum hardening in the fully aged state. Has the units of stress.

jvaara commented on src/DSA.jl:
```diff
@@ -1,174 +1,313 @@
+`b`: multiplier for yield strength evolution
+`w`: ???
+`P1`, `P2`: ???
+`m`: ???
```
Parameter controlling the characteristic diffusion time. Depends on the type of diffusion. The value of 1/3 is thought to represent pipe diffusion along dislocation lines. Another typical value is 2/3.

jvaara commented on src/DSA.jl:
```diff
@@ -1,174 +1,313 @@
+`w`: ???
+`P1`, `P2`: ???
+`m`: ???
+`m1`, `m2`: ???
```
The exponent of the power-law type static recovery of kinematic hardening (Controls secondary creep and constant slope relaxation of stresses over longer period of time) which is activated in higher temperatures. Higher values (>6-10) effectively deactivate static recovery whereas lower values (<5) activate.

jvaara commented on src/DSA.jl:
```diff
@@ -1,174 +1,313 @@
+`P1`, `P2`: ???
+`m`: ???
+`m1`, `m2`: ???
+`M1`, `M2`: ???
```
The normalizer of the power-law type static recovery of backstresses. Has the units of stress. Can be used to activate/deactivate static recovery. Deactivation with high values.

jvaara commented on src/DSA.jl:
```diff
@@ -1,174 +1,313 @@
+`plastic_strain`: plastic part of strain tensor
+`cumeq`: cumulative equivalent plastic strain (scalar, ≥ 0)
+`R`: yield strength
+`ta`: ???
```
Effective aging time

jvaara commented on src/DSA.jl:
```diff
@@ -1,174 +1,313 @@
+`cumeq`: cumulative equivalent plastic strain (scalar, ≥ 0)
+`R`: yield strength
+`ta`: ???
+`Ra`: ???
```
Aging stress

jvaara commented on src/DSA.jl:
```diff
@@ -1,174 +1,313 @@
+`m`: ???
+`m1`, `m2`: ???
+`M1`, `M2`: ???
+`ba`: ???
```
A dimensionless parameter that controls the rate of evolution of aging stress to its asymptotic value. (Similar to the isotropic hardening `b`)

jvaara commented on src/DSA.jl:
```diff
@@ -1,174 +1,313 @@
+`m1`, `m2`: ???
+`M1`, `M2`: ???
+`ba`: ???
+`xi`: ???
```
A dimensionless parameter that controls the magnitude of the Marquis effect from the aging stress (whether the increased hardening due to aging shows as increased relaxation or not). Support [0,1]. With the value of 0, the aging stress contributes solely to the size of the yield surface `R` (isotropic hardening). With the value of 1, the aging stress contributes solely to the viscoplastic drag stress `K`.

jvaara commented on src/memory.jl:
```diff
@@ -1,203 +1,318 @@
+`nn`: plasticity multiplier exponent
+`C1`, `D1`: parameters governing behavior of backstress X1
+`C2`, `D2`: parameters governing behavior of backstress X2
+`Q0`: ???
```
The initial isotropic hardening saturation value. Has the units of stress.

jvaara commented on src/memory.jl:
```diff
@@ -1,203 +1,318 @@
+`C1`, `D1`: parameters governing behavior of backstress X1
+`C2`, `D2`: parameters governing behavior of backstress X2
+`Q0`: ???
+`QM`: ???
```
The asymptotic isotropic hardening saturation value reached with high strain amplitude. Has the units of stress.

jvaara commented on src/memory.jl:
```diff
@@ -1,203 +1,318 @@
+`C2`, `D2`: parameters governing behavior of backstress X2
+`Q0`: ???
+`QM`: ???
+`mu`: ???
```
The parameter that controls the rate of evolution of the strain-amplitude dependent isotropic hardening saturation value.

jvaara commented on src/memory.jl:
```diff
@@ -1,203 +1,318 @@
+`Q0`: ???
+`QM`: ???
+`mu`: ???
+`b`: ???
```
The parameter that controls the rate of evolution for isotropic hardening.

jvaara commented on src/memory.jl:
```diff
@@ -1,203 +1,318 @@
+`QM`: ???
+`mu`: ???
+`b`: ???
+`eta`: ???
```
Parameter that controls the balance between memory surface kinematic and isotropic hardening. Support [0,1]. With the value of 0, the memory surface hardens kinematically; with the value of 1, the memory surface hardens isotropically. Initially values of 1/2 were used by several authors. Later values <1/2 have been suggested to capture the progressive process of memorization.

jvaara commented on src/memory.jl:
```diff
@@ -1,203 +1,318 @@
+end
+
+# TODO: complete this docstring
+"""Parameter state for Memory material.
```
Strain memory material. Strain memory is used to be able to model strain amplitude-dependent isotropic hardening. In practice, the transition from tensile test curve to cyclic behavior can be captured with this model.

jvaara commented on src/memory.jl:
```diff
@@ -1,203 +1,318 @@
+`plastic_strain`: plastic part of strain tensor
+`cumeq`: cumulative equivalent plastic strain (scalar, ≥ 0)
+`R`: yield strength
+`q`: ???
```
size of the strain memory surface (~plastic strain amplitude)

jvaara commented on src/memory.jl:
```diff
@@ -1,203 +1,318 @@
+`cumeq`: cumulative equivalent plastic strain (scalar, ≥ 0)
+`R`: yield strength
+`q`: ???
+`zeta`: ???
```
memory surface kinematic hardening variable

jvaara commented on src/memory.jl:
```diff
@@ -1,203 +1,318 @@
+    b::T = 0.0
+    eta::T = 0.0
+    m::T = 0.0
+    pt::T = 0.0
```
Threshold of equivalent plastic strain, after which the memory evanescence starts.

jvaara commented on src/memory.jl:
```diff
@@ -1,203 +1,318 @@
+    mu::T = 0.0
+    b::T = 0.0
+    eta::T = 0.0
+    m::T = 0.0
```
Memory evanescence exponent. Controls the non-linearity of the memory evanescence

jvaara commented on src/memory.jl:
```diff
@@ -1,203 +1,318 @@
+    eta::T = 0.0
+    m::T = 0.0
+    pt::T = 0.0
+    xi::T = 0.0
```
Memory evanescence strength multiplier

jvaara commented on src/DSA.jl:
```diff
@@ -1,174 +1,313 @@
+    xi::T = 0.0
+end
+
+# TODO: complete this docstring
```
A model that captures dynamic (and static) strain aging (DSA) induced hardening. The related phenomena are:
- Portevin le Chatelier effect (Serrated yield, plastic instabilities, see https://doi.org/10.1016/j.commatsci.2011.05.039)
- Discontinuous yielding
- Inverse strain rate sensitivity
- Secondary hardening in low cycle fatigue (LCF) tests
These typically occur in certain temperature/strain rate regime, where the dislocations are pinned due to diffusion of solute atoms. In the most effective conditions, the speed of diffusion is comparable to the applied strain rate (speed of dislocations).

jvaara commented:
Do you consider this finished so that it could be merged?

ahojukka5 commented:
https://github.com/JuliaFEM/Materials.jl/pull/66/commits

Looking at commit topics, some of those are not very descriptive. In the best case, every commit should have a descriptive topic, what it is doing, and doing only one clear thing. This is not a request for starting rebasing things, just note how things could be done better in the future.

https://github.com/erlang/otp/wiki/Writing-good-commit-messages

ahojukka5 commented:
One way to check how things are is to write `git log --pretty=oneline`. If the topics make sense, i.e. one can understand the outline what the commits are doing without actually using `git show <sha>` to see the changes, then everything is ok. But if there is commits like "WIP" or "Ops, changed back, not working", it means that some refactoring could be done. The basic tools are reword, fixup and squash. It is also possible to break one commit to several smaller commits afterwards. And it's also possible to make partial commits, so if there is a lot of changes in the same file such that they don't have anything common, it's better to make two or more separate commits.

Technologicat commented:
> Do you consider this finished so that it could be merged?

@jvaara: Yes. I left this dangling on purpose for a while, so that if anything else comes to mind, I could still add it. But it's been two weeks now, so yes, let's call this PR finished.

Technologicat commented:
@ahojukka5 :  Thanks for the git log review.

Generally speaking, I do aim for one thing per commit, with a descriptive message, but I haven't bothered to clean up the "oops" and "WIP" commits in a work in progress. Also, to my understanding, as a general rule, if a commit has been published, it's better not to change it. OTOH, this work was done in a branch that will be deleted when done, so I suppose that rule doesn't apply, until we merge.

As someone once said, Git is pretty good at [not having to say, "you should have"](https://tomayko.com/blog/2008/the-thing-about-git). Still, in my experience there have been cases where the changes to the same file are so tangled that cleanly separating them requires first undoing some of them. I'm not sure if that was the case here, though, or if it's just a case of me having been lazy.

Regarding breaking an existing commit into several, thanks for the tip, I didn't know that.

The `git log --pretty=oneline` tip is going into my `.bashrc`, thanks. Though nowadays I mostly interact with git through Emacs with [Magit](https://magit.vc/). In the status view, pressing `l l` (two lowercase ell) gives pretty much the same output as `git log --pretty=oneline`, plus you can highlight an entry and press `RET` to view it (as well as use any of Emacs's search packages such as `helm-swoop` to look for a specific thing in the commit titles). Splitting clean changes into several commits is trivial and very fast, just select the part of the diff you want and hit `s` to stage (or `u` to unstage). Interactive rebase is just `r i`, it'll open a view with single-key commands to squash, pick or whatever (with on-screen instructions). And you get to write the commit messages with Emacs. ;)

So, I suppose the final question is, do we want to clean the commit history before merging, or just merge as-is?

ahojukka5 commented:
If the commit history is not bothering you, then I would just merge as-is. But I encourage, when you have some spare time, to do some experiments with git rebase, it's actually very interesting to find out how powerful tool rebasing is.
